torch/export/export_schema.py

```diff
@@ -0,0 +1,326 @@
+        # SYMINT = auto()    # !!! Can Symint[] be an arguement? there are symint[] in native_function.yaml
+
+        # GRAPH = auto()      # !!! Consider how to handle condition op, which need to pass in a graph for the branch
+        # GRAPHS = auto()      # !!! What about list of graphs? Do we need this?
+
+        # !!! Following types doesn't have a list version in native_function.yaml
+        SCALAR_TYPE = auto()
+        MEMORY_FORMAT = auto()
+        LAYOUT = auto()
+        DEVICE = auto()
+
+    # Declaration of which type the value field is storing
+    type: ArgumentType
+
+    val: Union[
+        TensorArgument,
+        # List[TensorArgument],
+        Scalar,
+        # List[Scalar],
+        bool,
+        List[bool],
+        int,
+        List[int],
+        float,
+        List[float],
+        str,
+        # List[str],
+        # SymInt,
+        # List[SymInt],
+        # Graph,
+        # List[Graph],
+        ScalarType,
+        MemoryFormat,
+        Layout,
+        Device,
+    ]
+
+@dataclass
+class KeywordArgument:
+    key: str
+    vale: Argument
+
+# !!! How to model optional fields? Is it an operator schema annotation, or an argument type?
+#     Tensor?
+#     Scalar?
+#     ScalarType?
+#     bool?
+#     int?
+#     int[]?
+#     float[]?
+#     SymInt[]?
+#     MemoryFormat?
+#     Layout?
+#     Device?
+
+
+################################################################################
+# Following section is the defining the schema of serializing a concrete tensor
+
+# TensorMeta is a decription of a tensor, without the actual data (,effectively maps to FakeTensor)
+# TensorMeta has multliple uses
+#   1. Represent the property of a concrete tensor backed by a storage
+#     - This is used in the serialization of a concrete tensor, e.g. model weight
+#     - In this case, sizes and strides must be concrete ints, and cannot be symbolic
+#     - stride and storage_offset have to used to correctly reconstruct the tensor from the storage
+#   2. Represent the property of a virtual tensor (see IValue below)
+#     - In this case, sizes and strides can be either concrete ints or symbolic ints.
+#     - device/strides/storage_offset/layout/memory_format are tied to pytorch's implementation.
+#       These are faithful capture of tensor's detail in pytorch's executions during tracing
+#       However, it's up to downstream system on how to utilized these fields
+#       In another word, these feilds are suggestive, rather than mandatory.
+
+class TensorMeta:
+    dtype: ScalarType
+    sizes: List[SymInt]
+
+    # needed for training
+    requires_grad: bool
+
+    # !!! see description above, there are subtle difference on how these fields should be interpreted
+    device: Device
+    strides: List[SymInt]
+    storage_offset: int
+    layout: Layout
+    memory_format: MemoryFormat
+
+
+@dataclass
+class Buffer:
+    buffer: bytes
+
+
+@dataclass
+class ExternalBuffer:
```

Comment:
This seems too implementation-detail...

Reply:
For large models, model weights could be 10s of GB. 
It's a common requirement to store model weight in a separated file, that can be directly mmap into the memory. 

Also, for production scenarios, people wish to swap out model weight without touching model architecture file. There are benefits to decouple model file from model weights.

Reply:
I think we should be clear about the goals for this format, per the comment at the very top of the file.
Whatever we can do to make it more generally useful is great, but we should be intentional about what this should or should not support (i.e. do we really want to put production deployment as a goal here?)

Reply:
So, per the discussion today, the relevant goal here is for parameter (weight) storage to be portable and efficient.

Related, in mlir there was a similar proposal to move "weights" or basically constant values out of the .mlir file itself (some of an original design oversight there). https://discourse.llvm.org/t/rfc-adding-a-generalized-config-section-to-mlir-files/62867/39
That proposal leaned towards adding an abstraction for general resources or assets, of which parameter files were one of them.
I personally don't feel that's right, at least here, since this scheme is concerned with program representation, and anything beyond that (e.g. configurations, tooling assets, ins/outs used for export, etc.) should be a separate concern (I'd argue a wrapper object), particularly given the extra complication than an open serialization entry point would add here.

Anyway, just want to bring this up for completeness of the design.